store/store.go

```diff
@@ -261,24 +264,66 @@ func (s *Store) GetModel(name string) *Model {
 
 // Reduce processes txn events into the models.
 func (s *Store) Reduce(events []core.Event) error {
-	codecActions, err := s.eventcodec.Reduce(events, s.datastore, baseKey)
+	txn, err := s.datastore.NewTransaction(false)
 	if err != nil {
 		return err
 	}
-	actions := make([]Action, len(codecActions))
-	for i, ca := range codecActions {
+	defer txn.Discard()
+
+	actions := make([]Action, len(events))
+	for i, e := range events {
+		key := baseKey.ChildString(e.Model()).ChildString(e.EntityID().String())
 		var actionType ActionType
-		switch codecActions[i].Type {
+		var ca *core.CodecResult
+		switch e.Type() {
 		case core.Create:
 			actionType = ActionCreate
+			exist, err := txn.Has(key)
+			if err != nil {
+				return err
+			}
+			if exist {
+				return errCantCreateExistingInstance
+			}
+			ca, err = s.eventcodec.Reduce(e, nil)
```

Comment:
As a result, Reduce now runs on each individual Event, rather than the array. They all still happen within a transaction though, so that safety is maintained.